fastdtw/_fastdtw.pyx

```diff
@@ -0,0 +1,490 @@
+
+        Examples
+        --------
+        >>> import numpy as np
+        >>> import fastdtw
+        >>> x_1d = np.array([1, 2, 3, 4, 5.0])
+        >>> y_1d = np.array([2, 3, 4.0])
+        >>> fastdtw.dtw(x, y)
+        (2.0, [(0, 0), (1, 0), (2, 1), (3, 2), (4, 2)])
+
+        Optimization Considerations
+        ---------------------------
+        This function runs fastest if the following conditions are satisfied:
+            1) x and y are either 1 or 2d numpy arrays whose dtype is a
+               subtype of np.float
+            2) The dist input is a positive integer or None
+    '''
+
+    cdef int len_x, len_y, x_idx, y_idx, idx
+    cdef double cost
+    x, y = __prep_inputs(x, y, dist)
+    len_x, len_y = len(x), len(y)
+    idx = 0
+    cdef WindowElement we
+    cdef vector[WindowElement] window
+    window.resize(len_x * len_y)
+
+    for x_idx in range(len_x):
+        for y_idx in range(len_y):
+            we.x_idx = x_idx
+            we.y_idx = y_idx
+
+            if x_idx == y_idx == 0:
+                we.cost_idx_left = -1
+                we.cost_idx_up = -1
+                we.cost_idx_corner = 0
+            else:
+                we.cost_idx_left = -1 \
+                    if y_idx == 0 and x_idx > 0 \
+                    else idx
+                we.cost_idx_up = -1 \
+                    if x_idx == 0 and y_idx > 0 \
+                    else idx - len_y + 1
+                we.cost_idx_corner = -1 \
+                    if we.cost_idx_left == -1 or we.cost_idx_up == -1 \
+                    else we.cost_idx_up - 1
+
+                if we.cost_idx_left == 0:
+                    we.cost_idx_left = -1
+                if we.cost_idx_up == 0:
+                    we.cost_idx_up = 0
+
+            window[idx] = we
+            idx += 1
+
+    cdef PathElement *path = (  # noqa
+        <PathElement *>PyMem_Malloc(  # noqa
+            (len(x) + len(y) - 1) * sizeof(PathElement)))
+
+    cdef int path_len = 0
+    cost = __dtw(x, y, window, dist, path, path_len)
+
+    path_lst = []
+    if path != NULL:
+        path_lst = [(path[i].x_idx, path[i].y_idx)
+                    for i in range(path_len)]
+        PyMem_Free(path)
+
+    return cost, path_lst
+
+
+cdef inline double manhattan(double a, double b):
+    return fabs(a - b)
+
+
+def __create_linalg_norm(p):
+    def f(a, b):
+        return np.linalg.norm(a - b, p)
+
+    return f
+
+cdef __prep_inputs(x, y, dist):
+    if isinstance(dist, numbers.Number) and dist <= 0:
+        raise ValueError('dist cannot be a negative integer')
+    elif (isinstance(x, np.ndarray) and
+          isinstance(y, np.ndarray) and
+          x.ndim > 1 and y.ndim > 1 and
+          x.shape[1] != y.shape[1]):
+
+        raise ValueError('x and y must have the same width')
+
+    if isinstance(dist, numbers.Number):
+        # p-norm operations assume numpy array
+        if not isinstance(x, np.ndarray):
+            x = np.array(x)
```

Comment:
This can be concisely written as
```py
x = np.asanyarray(x)
```
If dtype=float is assumed later one,
```py
x = np.asanyarray(x, dtype=float)
```